src/bin/pg_autoctl/monitor.c

```diff
@@ -2031,6 +2115,247 @@ monitor_get_notifications(Monitor *monitor)
+			if (strcmp(notification.formationId, formation) != 0)
+			{
+				continue;
+			}
+
+			if (notification.reportedState == PRIMARY_STATE
+				&& notification.goalState == APPLY_SETTINGS_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 1/4: primary node %d (%s:%d) is assigned \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.goalState));
+			}
+			else if (notification.reportedState == APPLY_SETTINGS_STATE
+					 && notification.goalState == APPLY_SETTINGS_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 2/4: primary node %d (%s:%d) reported \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.reportedState));
+			}
+			else if (notification.reportedState == APPLY_SETTINGS_STATE
+					 && notification.goalState == PRIMARY_STATE)
+			{
+				applySettingsTransitionInProgress = true;
+
+				log_debug("step 3/4: primary node %d (%s:%d) is assigned \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.goalState));
+			}
+			else if (applySettingsTransitionInProgress
+					 && notification.reportedState == PRIMARY_STATE
+					 && notification.goalState == PRIMARY_STATE)
+			{
+				applySettingsTransitionDone = true;
+
+				log_debug("step 4/4: primary node %d (%s:%d) reported \"%s\"",
+						  notification.nodeId,
+						  notification.nodeName,
+						  notification.nodePort,
+						  NodeStateToString(notification.reportedState));
+			}
+
+			/* prepare next iteration */
+			PQfreemem(notify);
+			PQconsumeInput(connection);
+		}
+	}
+
+	/* disconnect from monitor */
+	pgsql_finish(&monitor->pgsql);
+
+	return applySettingsTransitionDone;
+}
+
+
+/*
+ * monitor_wait_until_node_reported_maintenance receives notifications and
+ * watches for the given node to have goalState and reportedState set to given
+ * state.
+ *
+ * If we lose the monitor connection while watching for the transition steps
+ * then we stop watching. It's a best effort attempt at having the CLI be
+ * useful for its user, the main one being the test suite.
+ */
+bool
+monitor_wait_until_node_reported_state(Monitor *monitor,
+									   int nodeId,
+									   NodeState state)
+{
+	PGconn *connection = monitor->pgsql.connection;
+	bool reachedMaintenance = false;
+
+	if (connection == NULL)
+	{
+ 		log_warn("Lost connection.");
+		return false;
+	}
+
+	while (!reachedMaintenance)
+	{
+		/* Sleep until something happens on the connection. */
+		int         sock;
+		fd_set      input_mask;
+		PGnotify   *notify;
+
+		sock = PQsocket(connection);
+
+		if (sock < 0)
+			return false;	/* shouldn't happen */
+
+		FD_ZERO(&input_mask);
+		FD_SET(sock, &input_mask);
+
+		if (select(sock + 1, &input_mask, NULL, NULL, NULL) < 0)
```

Comment:
socket  magic. Prefer if we can use PGxxx

Reply:
See above, that's per Postgres documentation. Now with a comment about it.